fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/rule/SingleTableRewriteBaseRule.java

```diff
@@ -45,9 +49,48 @@ public List<OptExpression> transform(OptExpression queryExpression, OptimizerCon
             for (OptExpression expression : expressions) {
                 calculateStatistics(expression, context);
             }
-            // sort expressions based on statistics output row count
-            expressions.sort(Comparator.comparingDouble(expression -> expression.getStatistics().getOutputRowCount()));
-            return expressions.subList(0, 1);
+            List<CandidateContext> contexts = Lists.newArrayList();
+            for (int i = 0; i < expressions.size(); i++) {
+                Statistics mvStatistics = collectMVStatistics(expressions.get(i));
+                Preconditions.checkState(mvStatistics != null);
+                contexts.add(new CandidateContext(mvStatistics, i));
+            }
+            // sort expressions based on statistics output row count and compute size
+            contexts.sort(new CandidateContextComparator());
+            return Lists.newArrayList(expressions.get(contexts.get(0).getIndex()));
+        }
+    }
+
+    @VisibleForTesting
+    public static class CandidateContext {
+        private Statistics mvStatistics;
+        private int index;
+
+        public CandidateContext(Statistics mvStatistics, int index) {
+            this.mvStatistics = mvStatistics;
+            this.index = index;
+        }
+
+        public Statistics getMvStatistics() {
+            return mvStatistics;
+        }
+
+        public int getIndex() {
+            return index;
+        }
+    }
+
+    @VisibleForTesting
+    public static class CandidateContextComparator implements Comparator<CandidateContext> {
+        @Override
+        public int compare(CandidateContext context1, CandidateContext context2) {
+            int ret = Double.compare(context1.getMvStatistics().getOutputRowCount(),
+                    context2.getMvStatistics().getOutputRowCount());
+            if (ret != 0) {
```

Comment:
This isn't right. for the query group by a, MV1: group by a,b, MV2:group by a, we should always choose MV2.

Reply:
there is an anti-example:

  1. count(1) group by a,b  is slightly larger than count(1) group by a;
  2. column a's distribution skews drastically, but column group (a,b) 's distribution is smooth.